File: leetcode/0141-linked-list-cycle/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        """
        RUNTIME: 66 ms (80.52%)
        MEMORY: 17.6 MB (68.37%)
        """
        if not head or not head.next or not head.next.next: return False
        slow, fast = head.next, head.next.next
        while slow != fast:
            if not fast.next or not fast.next.next:
                return False
            slow = slow.next
            fast = fast.next.next
        return True

File: leetcode/0141-linked-list-cycle/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * RUNTIME: 11 ms (90.91%)
     * MEMORY: 8 MB (82.46%)
    */
    bool hasCycle(ListNode *head) {
        if (!head || !head->next || !head->next->next) {
            return false;
        }
        ListNode* slow = head->next;
        ListNode* fast = head->next->next;
        while (slow != fast) {
            if (!fast->next || !fast->next->next) {
                return false;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        return true;
    }
};
